FirebaseFunctions/Sources/FunctionsError.swift

```diff
@@ -207,53 +194,53 @@ extension FunctionsErrorCode {
                    code: rawValue,
                    userInfo: userInfo ?? [NSLocalizedDescriptionKey: descriptionForErrorCode])
   }
-}
 
-func FunctionsErrorForResponse(status: NSInteger,
+  static func errorForResponse(status: Int,
                                body: Data?,
                                serializer: FunctionsSerializer) -> NSError? {
-  // Start with reasonable defaults from the status code.
-  var code = FunctionsCodeForHTTPStatus(status)
-  var description = code.descriptionForErrorCode
-
-  var details: AnyObject?
+    // Start with reasonable defaults from the status code.
+    var code = FunctionsErrorCode.errorCode(forHTTPStatus: status)
+    var description = code.descriptionForErrorCode
+    var details: AnyObject?
+
+    // Then look through the body for explicit details.
+    if let body,
+       let json = try? JSONSerialization.jsonObject(with: body) as? NSDictionary,
+       let errorDetails = json["error"] as? NSDictionary {
+      if let status = errorDetails["status"] as? String {
+        code = .errorCode(forName: status)
+
+        // If the code in the body is invalid, treat the whole response as malformed.
+        guard code != .internal else {
+          return code.generatedError(userInfo: nil)
+        }
+      }
 
-  // Then look through the body for explicit details.
-  if let body,
-     let json = try? JSONSerialization.jsonObject(with: body) as? NSDictionary,
-     let errorDetails = json["error"] as? NSDictionary {
-    if let status = errorDetails["status"] as? String {
-      code = FunctionsErrorCode.errorCode(forName: status)
+      if let message = errorDetails["message"] as? String {
+        description = message
+      } else {
+        description = code.descriptionForErrorCode
+      }
 
-      // If the code in the body is invalid, treat the whole response as malformed.
-      guard code != .internal else {
-        return code.generatedError(userInfo: nil)
+      // Update `details` only if the object can be decoded.
+      if let innerDetails = errorDetails["details"],
+         let decodedDetails = try? serializer.decode(innerDetails) {
+        details = decodedDetails
       }
```

Comment:
I updated this part to make its behavior more apparent, without assigning `errorDetails["details"]` first and then potentially erasing it with `try?`. This matches the way it currently works:
```swift
details = errorDetails["details"] as AnyObject?
if let innerDetails = details {
  // Just ignore the details if there an error decoding them.
  details = try? serializer.decode(innerDetails)
}
```

However, I’m not quite sure I understood the original intention. If decoding fails, should `details` be `nil` or keep the undecodable object? Should it be something like this instead?
```swift
details = errorDetails["details"] as AnyObject?
// Update `details` only if decoding succeeds;
// otherwise, keep the original object.
if let innerDetails = details,
    let decodedDetails = try? serializer.decode(innerDetails) {
  details = decodedDetails
}
```

Reply:
> However, I’m not quite sure I understood the original intention. If decoding fails, should `details` be `nil` or keep the undecodable object? Should it be something like this instead?
> 
> ```swift
> details = errorDetails["details"] as AnyObject?
> // Update `details` only if decoding succeeds;
> // otherwise, keep the original object.
> if let innerDetails = details,
>     let decodedDetails = try? serializer.decode(innerDetails) {
>   details = decodedDetails
> }
> ```

I think this approach makes sense since it allows for falling back to the undecodable object, if it's non-nil.

Reply:
Alright, I updated this code (though it changes the method’s current behavior).

Should probably add tests in the future to set clear expectations.